src/operators/join.rs

```diff
@@ -320,100 +320,127 @@ impl<G, T1> JoinCore<G, T1::Key, T1::Val, T1::R> for Arranged<G,T1>
                         for batch2 in input2_buffer.drain(..) {
-                            if !batch2.is_empty() {
-                                if let Some(acknowledged1) = &acknowledged1 {
-                                    // TODO : cursor_through may be problematic for pre-merged traces.
-                                    // A trace should provide the contract that whatever its `set_physical_compaction` capability,
-                                    // it is safe (and reasonable) to await delivery of batches up through that frontier.
-                                    // In this case, we should be able to await (not block on) the arrival of these batches.
+                            // Ignore any pre-loaded data.
+                            if PartialOrder::less_equal(&acknowledged2, &batch2.lower()) {
+                                if !batch2.is_empty() {
+                                    // It is safe to ask for `ack1` as we validated that it was at least `get_physical_compaction()`
+                                    // at start-up, and have held back physical compaction ever since.
                                     let (trace1_cursor, trace1_storage) = trace1.cursor_through(acknowledged1.borrow()).unwrap();
                                     let batch2_cursor = batch2.cursor();
-                                    todo2.push_back(Deferred::new(trace1_cursor, trace1_storage, batch2_cursor, batch2.clone(), capability.clone(), |r1,r2| (r1.clone()) * (r2.clone())));
-                                }
-                            }
-                            // It would be alarming (incorrect) to receieve a non-empty batch that does not advance the
-                            // acknowledged frontier, as each batch must be greater than previous batches, and the input.
-                            // Empty batches may be received as information races between frontier forwarding of the trace
-                            // and the empty batches themselves (which can be sent as part of trace importing).
-                            if acknowledged2.is_none() { acknowledged2 = Some(timely::progress::frontier::Antichain::from_elem(<G::Timestamp>::minimum())); }
-                            if let Some(acknowledged2) = &mut acknowledged2 {
-                                if !PartialOrder::less_equal(&*acknowledged2, batch2.upper()) {
-                                    if !batch2.is_empty() {
-                                        panic!("Non-empty batch2 upper not beyond acknowledged frontier: {:?}, {:?}", batch2.upper(), acknowledged2);
-                                    }
+                                    todo2.push_back(Deferred::new(trace1_cursor, trace1_storage, batch2_cursor, batch2.clone(), capability.clone()));
                                 }
+
+                                // To update `acknowledged1` we might presume that `batch1.lower` should equal it, but we
```

Comment:
I think this should be `acknowledged2` and `batch2`

Reply:
Yes, oops.